### PR TITLE
GLA-731: Upgrade Luniverse-Gluwacoin to solidity 0.8

### PR DESCRIPTION
**Pull request by marc-aurele-besner**
## Notable changes:
- Upgrade contracts to Solidity 0.8.12
- 100% coverage test
- Deployment scripts to deploy one or all contracts;
- CLI UI to easily run all deployment, test, and more;
- Log of all deployed contracts per network;
- Move most of the config in hardhat.config.js file
- Move RPC and private key in .env.development
- Tool in cli to setup environment variables (private key and rpc)

## CLI UI usage
`npm run cli`

### Main options
- Setup all environment variables (rpc + private key)
- Run tests (select a test and a network, if not hardhat, it will use contract address from contractsAddressDeployed.json (if existent))
- Deploy contracts (select a contracts or all and a network)
- Deploy all contracts and run the tests sele (select a test and a network)
- Get account balance (Select a network)
- Get the previously deployed contract's address (if contractsAddressDeployed.json exist)
- Get all the previously deployed contracts address (if contractsAddressDeployedHistory.json exist)
- contractsAddressDeployed.json keep the last contracts of each network that you deployed (no duplicate)
- contractsAddressDeployedHistory.json keep ALL contracts deployed

### Coverage Test Result
Coverage Test Result

### CODE REVIEW

File: contracts/abstracts/ERC20Pausable.sol [hunk 1]
```diff
@@ -13,15 +12,23 @@ import "./BeforeTransferERC20.sol";
  * period, or having an emergency switch for freezing all token transfers in the
  * event of a large bug.
  */
-contract ERC20Pausable is BeforeTransferERC20, Pausable {
+contract ERC20Pausable is BeforeTransferERC20, PausableUpgradeable {
```
Comment (kevinnguyen17) [inline contracts/abstracts/ERC20Pausable.sol:15]: We do not need this contract anymore, just use the PausableUpgradeable

File: contracts/abstracts/Reservable.sol [hunk 2]
```diff
@@ -70,21 +71,27 @@ contract Reservable is BeforeTransferERC20 {
...
-        bytes32 hash = keccak256(abi.encodePacked(address(this), sender, recipient, executor, amount, fee, nonce, expiryBlockNum));
+        uint256 chainId = chainId();
+        bytes32 hash = keccak256(abi.encodePacked(GluwacoinModels.SigDomain.Reserve, chainId, address(this), sender, recipient, executor, amount, fee, nonce, expiryBlockNum));
         Validate.validateSignature(hash, sender, sig);
 
-        _reserved[sender][nonce] = Reservation(amount, fee, recipient, executor, expiryBlockNum,
-            ReservationStatus.Active);
-        _totalReserved[sender] = _totalReserved[sender].add(total);
+        _reserved[sender][nonce] = Reservation(amount, fee, recipient, executor, expiryBlockNum, ReservationStatus.Active);
+        unchecked {
```
Comment (kevinnguyen17) [inline contracts/abstracts/Reservable.sol:92]: This one can't use unchecked, we must remove the unchecked so solidity will revert if there is overlow

**Review by kevinnguyen17 — CHANGES_REQUESTED** [review]

**Review by kevinnguyen17 — COMMENTED** [review]
should add in pragma ^ so that all the minor versions of 0.8.0 will be supported.

File: contracts/abstracts/ETHlessTransfer.sol [hunk 3]
```diff
@@ -43,10 +39,9 @@ contract ETHlessTransfer is ContextUpgradeable, BeforeTransferERC20, GluwaRole {
      */
     function transfer(address sender, address recipient, uint256 amount, uint256 fee, uint256 nonce, bytes memory sig)
     public onlyGluwa returns (bool success) {
-        _useNonce(sender, nonce);
+        _useNonce(sender, 3, nonce); // 3 - Transfer
```
Comment (kevinnguyen17) [inline contracts/abstracts/ETHlessTransfer.sol:42]: Must use sigDomain.Transfer

File: contracts/LuniverseGluwacoin.sol [hunk 4]
```diff
@@ -1,36 +1,101 @@
...
+        uint256 fee,
+        uint256 nonce,
+        bytes calldata sig
+    ) external whenNotPaused {
+        uint256 burnerBalance = balanceOf(burner);
+        require(
+            burnerBalance >= amount,
+            "ERC20Wrapper: burn amount exceed balance"
+        );
+        _useNonce(burner, 1, nonce); // 1 - Burn
```
Comment (kevinnguyen17) [inline contracts/LuniverseGluwacoin.sol:78]: must use SigDomain

File: contracts/LuniverseGluwacoin.sol [hunk 5]
```diff
@@ -1,36 +1,101 @@
...
+
+    function _beforeTokenTransfer(address from, address to, uint256 amount) override(Reservable, ERC20Upgradeable) internal {
+        require(!super.paused(), "ERC20Pausable: token transfer while paused");
+
+        return super._beforeTokenTransfer(from, to, amount);
+    }
+
+    function reserve(address sender, address recipient, address executor, uint256 amount, uint256 fee, uint256 nonce,
+        uint256 expiryBlockNum, bytes memory sig) override public whenNotPaused returns (bool success) {
+        _useNonce(sender, 4, nonce); // 4 - Reserve
```
Comment (kevinnguyen17) [inline contracts/LuniverseGluwacoin.sol:62]: must use SigDomain

File: contracts/abstracts/BeforeTransferERC20.sol [hunk 6]
```diff
@@ -1,269 +1,55 @@
...
-     * - `owner` cannot be the zero address.
-     * - `spender` cannot be the zero address.
-     */
-    function _approve(address owner, address spender, uint256 amount) internal {
-        require(owner != address(0), "ERC20: approve from the zero address");
-        require(spender != address(0), "ERC20: approve to the zero address");
-
-        _allowances[owner][spender] = amount;
-        emit Approval(owner, spender, amount);
+    function _useNonce(address signer, uint8 domain, uint256 nonce) internal {
```
Comment (kevinnguyen17) [inline contracts/abstracts/BeforeTransferERC20.sol:50]: can change the uint8 domain to SigDomain to enhance readability

**Review by kevinnguyen17 — CHANGES_REQUESTED** [review]

**Review by kevinnguyen17 — APPROVED** [review]

**Review by ohyung1287 — APPROVED** [review]

**Pull request merged**